azuredevops.go

```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
+	return gitURL, nil
 }
 
-func buildURL(urlStr string, group string, project string) (string, error) {
-	urlPath, err := url.Parse(urlStr)
+func getProjectsWritesError(c *gin.Context, i importBody) (azureDevOpsProjectResponse, error) {
+	var getProjectsURL *url.URL
+	var format string
+	var values []interface{}
+	if i.Project != "" {
+		format = "%v/%v/%v"
+		values = []interface{}{ i.Group, apiProjects, i.Project }
+	} else {
+		format = "%v/%v"
+		values = []interface{}{ i.Group, apiProjects }
+	}
+
+	const apiVersion string = "5.0"
+	getProjectsURL, err := requests.ConstructGetURL(i.URL, map[string][]string{
+		"api-version": {apiVersion},
+	}, format, values...)
+
 	if err != nil {
-		fmt.Println("Unable to get url: ", err)
-		return "", err
+		var errorDetail string
+
+		if i.Project != "" {
+			errorDetail = fmt.Sprintf("Unable to build url '%v' for '%v/%v/%v'",
+				i.URL, values[0], values[1], values[2])
+		} else {
+			errorDetail = fmt.Sprintf("Unable to build url '%v' for '%v/%v'",
+				i.URL, values[0], values[1])
+		}
+
+		ginutil.WriteInvalidParamError(c, err, "URL", errorDetail)
+		return azureDevOpsProjectResponse{}, err
 	}
 
-	data := url.Values{}
-	data.Set("api-version", "5.0")
+	projects := azureDevOpsProjectResponse{
+		Count: 1,
+		Value: make([]azureDevOpsProject, 1),
+	}
 
-	if project != "" {
-		urlPath.Path = fmt.Sprintf("%v/%v/%v", group, apiProjects, project)
+	if i.Project != "" {
+		err = requests.GetAndParseJSON(&projects.Value[0], i.User, i.Token, getProjectsURL)
 	} else {
-		urlPath.Path = fmt.Sprintf("%v/%v", group, apiProjects)
+		err = requests.GetAndParseJSON(&projects, i.User, i.Token, getProjectsURL)
 	}
 
-	urlPath.RawQuery = data.Encode()
-	fmt.Println(urlPath.String())
+	if err != nil {
+		ginutil.WriteProviderResponseError(c, err, "Could be caused by invalid JSON data structure." +
+			"\nMight be the result of an incompatible version of Azure DevOps.")
+		return azureDevOpsProjectResponse{}, err
+	}
 
-	return urlPath.String(), nil
+	return projects, nil
 }
+
+func getProjectBranchesWritesError(c *gin.Context, i importBody, project string) ([]azureDevOpsBranch, error) {
+	const apiVersion string = "5.0"
+
+	urlPath, err := requests.ConstructGetURL(i.URL, map[string][]string{
+		"api-version": {apiVersion},
+		"filter": {"heads/"},
+	}, "%v/%v/%v/%v/%v", i.Group, project, apiRepositories, project, refsPath)
+
+	if err != nil {
+		ginutil.WriteInvalidParamError(c, err, "URL", fmt.Sprintf("Unable to parse URL %q", i.URL))
+		return []azureDevOpsBranch{}, err
+	}
+
+	fmt.Println(urlPath.String())
+
+	projectRefs := struct {
+		Value []azureDevOpsRef `json:"value"`
+		Count int              `json:"count"`
+	}{}
+
+	err = requests.GetAndParseJSON(&projectRefs, i.User, i.Token, urlPath)
+	if err != nil {
+		ginutil.WriteProviderResponseError(c, err, "Could be caused by invalid JSON data structure." +
+			"\nMight be the result of an incompatible version of Azure DevOps.")
```

Comment:
Suggest specifying what action caused this problem in the details. And suggest removing the newline as no other problem detail has those. It will be just as fine with just a space between them instead.